### ISSUE TITLE
A need for importer hooks

### ISSUE THREAD

**Issue by almasaeed2010** [tripal/tripal#553]
## Feature Request/Discussion

- [ ] does this feature attempt to solve an existing problem with Tripal?
- [x] are you open to developing or collaborating on an extension module if this is not a good fit for Tripal Core (no pressure here -just good to know upfront :-) )
- [ ] do you **Need** this feature ASAP?

### Description

We have encountered multiple cases where we needed to know that an importer had finished running. Is it possible to add a hook to the main `TripalImporter` class that gets triggered when the importer is done?

### Your Specific Use Case

In ElasticSearch, we need to know when annotation importers have finished to trigger a new indexing job. Since annotation importers such as interpro don't necessarily publish anything to the drupal side of things, we can't rely on field and entity hooks to know if we should trigger indexer jobs. 

It would also be helpful if the importer arguments are passed to the hook too which would allows us to determine which organism to reindex.

### Generally Applicable

This would also be needed for [tripal_cv_xray](https://github.com/statonlab/tripal_cv_xray), which is a cv term browser for tripal 3. Since it relies heavily on caching cv to entity relationships to be able to display the tree, it'd be extremely useful to run auto caching scripts when an importer is finished.

Any thoughts about this? Is there another way to do this without adding hooks to importers directly?

Thanks!

**Comment by spficklin** [comment]
Hi @almasaeed2010 . I think this is a very reasonable request and adds very minimal code.  I have added two new hooks:  hook_importer_start and hook_importer_finish.  These get called before and after a TripalImporter is run.   Can you test out the new branch named 553-tv3-importer_hooks and see if it works for you?  If so, we can issue a pull request.

**Comment by almasaeed2010** [comment]
Hi @spficklin 

I took a look at the code and I think that's exactly what needed. This will work for both ES and tripal_cv_xray.

My only suggestion would be to add a public method to TripalImporter to get the arguments of the loader. This is useful to identify the parts of the site that would need reindexing. 

To clarify, this is all we'd need:
```php
public function getArguments() {
  return $this->arguments;
}
```

Would that work?

@bradfordcondon any thoughts since you are maintaining tripal_cv_xray?

Thanks @spficklin!

**Comment by bradfordcondon** [comment]
I can see some utility getting, say, the organism or analysis id from the arguments so that would be helpful.

**Comment by spficklin** [comment]
Yes, I think that's reasonable.  My focus at the moment are the release blockers for the Tripal v3 stable release.  If you need this sooner rather than later please feel free to

**Comment by spficklin** [comment]
Oops... didn't mean to close.  Yes, I agree that's a reasonable request.

**Comment by almasaeed2010** [comment]
Just a note that this is not urgent but we don't mind picking up from here and adding the rest of the code along with some unit testing if you want to focus on more urgent issues.

Thanks!

**Comment by spficklin** [comment]
This was very easy to do, so I've added the getArguments function for you.  Just pull the branch and you'll have it.  If you need this to be included in Tv3 stable so things work for you then let's bump this up to a release blocker and create a PR.

**Comment by spficklin** [comment]
Oh, also.... as this started out as a feature request we should at least get the thumbs up from the other members of the PMC @shawnawsu and @laceysanderson .... thoughts?

**Comment by laceysanderson** [comment]
I'm fully in support of this feature :-) 👍

**Comment by almasaeed2010** [comment]
Implemented in #601
